### ISSUE TITLE
Serious data corruption bug

### ISSUE THREAD

**Issue by gramster** [notable/notable#603]
- **OS Version**: MacOS 10.14.4
- **Notable Version**: 1.4.0

### Current behavior

Okay, I just had a serious issue. I've had data corruption issues in the past but they didn't seem too bad (although now I wonder if I have lost notes I didn't even know I was losing). Anyway, due to the past issues, I recently added git to my notes directory. Unfortunately I hadn't committed these recent changes so I have lost some notes.

So what just happened: I had a notebook with 15 notes - a top level one, which linked to 14 subnotes. I clicked on one of the links to open the subnote, which at this point had nothing but a title. I hit enter to start typing a body, and the title disappeared (became "#"). At the same time (I believe), six of the other notes disappeared too! It's possible these weren't *exactly* correlated in time, but I went pretty quickly from opening Notability, opening the notebook and seeing all the subnotes, opening the top level one, clicking on the link to open a subnote, pressing ENTER, and then noticing a bunch of subnotes disappeared.

### Expected behavior

Don't delete notes, especially ones that aren't even in the process of being edited!

**Comment by gramster** [comment]
It looks like I can repro the issue with the title. It may even be by design. When I follow a note link, and then edit the linked-to note, the note opens in the editor with the title highlighted so pressing any key will replace the whole title with that key. I hit enter, and so ended up with a blank title. This is bad, but at least the second time it didn't result in other notebooks being destroyed.

**Comment by fabiospampinato** [comment]
Let me see if I've got the chain of events right:

1. You have 15 notes in a notebook
2. You clicked one of them
3. You're in preview mode (I suppose, otherwise you couldn't click a link in that note)
4. You click a link in that note that points to another note
5. Now that other note is the active one
6. You switch to edit mode (I suppose, otherwise you wouldn't be able to edit it)
7. The title gets selected (I suppose)
8. You press enter and the note's title gets removed
9. Multiple notes get deleted accidentally

Is that right?

If so I see two potential problems here:

1. A note's title is being selected when it shouldn't be: the way it works currently is that a note's title will be selected automatically when that note has no other content. The idea behind this is that perhaps you created a new note, whose initial content might be "# Untitled" and you'd generally want to change its title right away, so we preselect it. So was the note whose title got selected for you had no content? I suppose we may want to make this check more strict, ensuring that we are only selecting the title if it begins with "Untitled" 🤔 

2. Multiple notes got deleted accidentally: we are only deleting notes in 2 places (you can check this yourself by searching the codebase for "File.unlink"):
  1. [Here](https://github.com/notable/notable/blob/9817109d22d20a45532d10e08ddbd6216cbef12f/src/renderer/containers/main/note.ts#L194), if you tell the app to permanently delete a note we delete it from the disk.
  2. [Here](https://github.com/notable/notable/blob/9817109d22d20a45532d10e08ddbd6216cbef12f/src/renderer/containers/main/note.ts#L791) if a note needs to be renamed to something else (because it has been edited and it's title changed), and that something else is already present on disk, we delete the current note and overwrite the other one instead of what we usually do: renaming the current note and then overwriting it. This line of code should basically never get hit, the only way it could get executed is if perhaps there's a race condition with another program that is creating the other file after we check for its existence but before we can rename to it. Still I don't see how this could cause the deletion of multiple notes in any way.

The only way this could perhaps happen is if our [filesystem watcher](https://github.com/notable/notable/blob/master/src%2Frenderer%2Futils%2Fwatcher.ts) is detecting some rename events that aren't actually happening. In that case I think notes would be deleted right after the app started. If you can reproduce this issue reliably, can you check what events are being detecting by our filesystem watcher by putting a `console.log ( event, args )` [here](https://github.com/notable/notable/blob/9817109d22d20a45532d10e08ddbd6216cbef12f/src/renderer/utils/watcher.ts#L89) and checking the output in the console?

**Comment by RenanWRS** [comment]
This issue happened to me too, luckly I was abble to recover those files through Google Drive.

- **OS Version:** Windows 10
- **Notable Version:** 1.4.0

I don't know what I did to cause this problem, because it took me a while to realize that the files were missing, but what I know for sure is that I didn't click on the 'move to the trash' button.

27 notes were removed in total, of different tags, remaining only 20 notes, I couldn't find anything in common among all the notes removed.

**Comment by fabiospampinato** [comment]
There are 2 parts to this issues:

1. The title was getting highlighted when the note had no other content, this has been fixed and now it only happens when the title is actually "Untitled" or "Untitled (number)".

2. Notes get deleted automatically for some reason. @renanwillian mentioned he was using Google Drive, were you @gramster using any third-party synchronization service? If so then this might be the same issue as #515, which I'm pretty sure I've found why it's happening and how to fix it, it'd be extremely useful if you guys could try out v1.8 (to be released soon) and see if you can reproduce the issue.

**Comment by fabiospampinato** [comment]
I think this was a duplicate of #515, which is fixed now. If anybody can reproduce this or any other issue involving data loss please report it.

**Comment by gramster** [comment]
Sorry for the delay in answering; I was using iCloud for synchronization.

**Comment by smdahmed** [comment]
I am on version 1.8.4 and I use Mega Sync for file sync. I can see that overnight without a particular reason, a particular tag of my notes completely disappeared. They are replaced with Untitled and empty notes.

I use Notable on Ubuntu.

**Comment by fabiospampinato** [comment]
@smdahmed what bundle of the app did you install, the snap?

**Comment by smdahmed** [comment]
@fabiospampinato - I am just realising that I have installed through the app store and there is a known issue over there. I will reinstall. Is there any hope that I can recover my corrupted files?

**Comment by fabiospampinato** [comment]
@smdahmed To the best of my knowledge the strict confinement the snap was published with allowed the app to create new files but not to write to them afterwards, I didn't see this limitation coming at that time as everything was working within the app and files were being created on disk, I assumed the snap environment would have allowed Notable to edit files it itself created, but that's not the case for some reason.

Therefore unfortunately I think whatever you wrote in that notes never got saved in the first place. I'm sorry about this.

--

I've just now published a working snap to the store finally, it uses the "classic" confinement, which enables it to work properly. I had to wait a while until somebody on the snap team manually allowed me to use the classic confinement in the app published in the store.

**Comment by smdahmed** [comment]
@fabiospampinato - No worries at all. Thank you so much for correcting this. I must tell you, Notable is an app that I cant live without :). So kudos for all your hardwork. 

First point is that it is my mistake - I didnt read the description in app store that Notable is not working from the snap installation. 

All the best and please keep improving the app.
